[addons][screensaver] cleanup and improve screensaver code

AlwinEsch opened the pull request:
## Description

This cleanups the screensaver system to have more usable on planed Sandbox and more match current style.
Also becomes inside Kodi a `std::unique_ptr` used to have more safe.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch committed:
[addons][screensaver] take addon pointer directly to calls

Before that, the complete "C" structure pointer was passed, which is actually not necessary (and also unnecessary work to get pointers).
This changes it to only pass the pointer required by the add-on.

AlwinEsch committed:
[addons][screensaver] Used typedef to define "C" function pointer

This has the background to be able to use several parts in the sandbox with automatic creation.
It also makes the structure look cleaner.

In the commit, the value names are also changed to lowercase, which corresponds more to the "C" style.

AlwinEsch committed:
[addons][screensaver] Clean up "C" header

It removes once the doc lines, which are not really usable.
It also shifts the callback structure downwards.
Everything from line 50 would no longer be necessary in the sandbox and the callback would be there as:
```cpp
ATTR_DLL_EXPORT void kodi_addon_screensaver_get_properties(KODI_OWN_HDL hdl, struct SCREENSAVER_PROPS* props) __INTRODUCED_IN_KODI(1);
```

AlwinEsch committed:
[addons][screensaver] increase ADDON_INSTANCE_VERSION_SCREENSAVER[_MIN] to 2.2.0

garbear commented on xbmc/addons/ScreenSaver.cpp:
```diff
@@ -53,27 +45,40 @@ CScreenSaver::~CScreenSaver()
+  props->height = CServiceBroker::GetWinSystem()->GetGfxContext().GetHeight();
+  props->pixelRatio = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().fPixelRatio;
+  props->name = strdup(Name().c_str());
+  props->presets = strdup(CSpecialProtocol::TranslatePath(Path()).c_str());
```
`std::strdup` via &lt;cstring>. Also what I've done in the past, is add a std::string member, and assign its .c_str() to the variable, so that the data is cleaned via RAII

AlwinEsch replied:
Know what you mean, but can not works in all cases.
Reasons:
- In case a addon is in "C", has several instances and stores somewhere his pointer. Then can not be sure that pointer still exists.
  - in a good independent "C" should a `const char*` only be given if it stays in all cases present.
- Another part is about Sandbox where the addon runs independent from Kodi. There can never be a `const char*` given to another target.
  - As the sandbox addon library can not store then itself e.g. a `std::string` to get his "C" characters must be always a string created and then from addon deleted.
  - **Note:** By Sandbox inside function values can be a `const char*` given (as memory then in shared part) only on structures not so easy.

AlwinEsch replied:
Think to remove the three string values.
They are not really needed and can be also done with other callbacks (e.g. `kodi::addon::GetAddonPath()`).

AlwinEsch commented:
Have removed the "string" values, this values are on all screensaver addons not used and then also good to safe mem with remove.

AlwinEsch committed:
[addons][screensaver] remove give of string on properties

The given strings can also be get by other callbacks:
- .name via kodi::addon::GetAddonInfo("name");
- .presets via kodi::addon::GetAddonPath();
- .profile via kodi::addon::GetUserPath();

ksooo commented on xbmc/addons/ScreenSaver.cpp:
```diff
@@ -13,33 +13,25 @@
+
+namespace
+{
+static void get_properties(const KODI_HANDLE hdl, struct KODI_ADDON_SCREENSAVER_PROPS* props)
```
No need for `static` as this function is in an unnamed namespace, which has exactly the same effect and purpose, to reduce function's visibility to the compile unit.

ksooo commented on xbmc/addons/ScreenSaver.cpp:
```diff
@@ -13,33 +13,25 @@
+  if (hdl)
+    static_cast<CScreenSaver*>(hdl)->GetProperties(props);
+}
+} /* namespace */
```
c++ comment, please

ksooo commented on xbmc/addons/ScreenSaver.cpp:
```diff
@@ -53,27 +45,37 @@ CScreenSaver::~CScreenSaver()
+  props->device = CServiceBroker::GetWinSystem()->GetHWContext();
+  props->width = CServiceBroker::GetWinSystem()->GetGfxContext().GetWidth();
+  props->height = CServiceBroker::GetWinSystem()->GetGfxContext().GetHeight();
+  props->pixelRatio = CServiceBroker::GetWinSystem()->GetGfxContext().GetResInfo().fPixelRatio;
```
Only call `CServiceBroker::GetWinSystem()` once, assign to a variable and reuse that var.

AlwinEsch replied:
Thanks for the comments, also added now a pointer check to it (was also requested for other places by my font cleanup request).

ksooo commented on xbmc/windows/GUIWindowScreensaver.h:
```diff
@@ -10,7 +10,13 @@
+namespace ADDONS
+{
+class CScreenSaver;
+} /* namespace ADDONS */
```
c++ comment

ksooo commented on xbmc/windows/GUIWindowScreensaver.h:
```diff
@@ -10,7 +10,13 @@
+{
+class CScreenSaver;
+} /* namespace ADDONS */
+} /* namespace KODI */
```
c++ comment

ksooo commented on xbmc/windows/GUIWindowScreensaver.h:
```diff
@@ -26,6 +32,6 @@ class CGUIWindowScreensaver : public CGUIWindow
 
 private:
-  ADDON::CScreenSaver* m_addon = nullptr;
+  std::unique_ptr<KODI::ADDONS::CScreenSaver> m_addon;
```
`#include <memory>`missing in this header.

ksooo reviewed (changes requested):
Minors only.

AlwinEsch committed:
[addons][screensaver] Replaced AddonProps_Screensaver with callback function

The background here is also a planned sandbox system. Because the properties must be available first, it is difficult to have them defined before the add-on was created. Therefore it is hereby replaced with a callback.
In the C ++ header the previous way remains, once it calls the new "C" function to get the data.

AlwinEsch committed:
[addons][screensaver] change namespace from ADDON to KODI::ADDONS

To be more correct to Kodi's current namespace style.

AlwinEsch committed:
[addons][screensaver] use std::unique_ptr about CScreenSaver

This uses a unique_ptr pointer to make it safer to use, since "m_addon" may not be deleted under all circumstances.

ksooo reviewed (approved):
LGTM

AlwinEsch commented:
Thanks a lot @ksooo to look and approve. 
Want to bring in before the addons becomes updated, Then another version increase is not necessary at 11 addons :smirk::face_exhaling:

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
jenkins build this please
